test/pkg/environment/azure/byok.go

```diff
@@ -0,0 +1,262 @@
+		},
+	}
+
+	poller, err := env.keyVaultClient.BeginCreateOrUpdate(ctx, env.NodeResourceGroup, keyVaultName, armkeyvault.VaultCreateOrUpdateParameters{
+		Location:   keyVault.Location,
+		Properties: keyVault.Properties,
+	}, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	kvResp, err := poller.PollUntilDone(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &kvResp.Vault, nil
+}
+
+// assignKeyVaultRBAC assigns necessary RBAC roles for Key Vault access
+func (env *Environment) assignKeyVaultRBAC(ctx context.Context, keyVaultID, karpenterIdentity, testUserPrincipalID string) error {
+	cryptoOfficerRoleDefinitionID := fmt.Sprintf("/subscriptions/%s/providers/Microsoft.Authorization/roleDefinitions/%s", env.SubscriptionID, keyVaultCryptoOfficer)
+	keyVaultAdminRoleDefinitionID := fmt.Sprintf("/subscriptions/%s/providers/Microsoft.Authorization/roleDefinitions/%s", env.SubscriptionID, keyVaultAdminRole)
+
+	// Assign roles to karpenter workload identity so that karpenter can access keys for disk encryption
+	err := env.RBACManager.EnsureRole(ctx, keyVaultID, cryptoOfficerRoleDefinitionID, karpenterIdentity)
+	if err != nil {
+		return fmt.Errorf("failed to assign Crypto Officer role to karpenter identity: %w", err)
+	}
+	err = env.RBACManager.EnsureRole(ctx, keyVaultID, keyVaultAdminRoleDefinitionID, karpenterIdentity)
+	if err != nil {
+		return fmt.Errorf("failed to assign Administrator role to karpenter identity: %w", err)
+	}
+
+	// User from az.DefaultCred needs rbac to create keys and manage the vault (wrap, unwrap etc)
+	err = env.RBACManager.EnsureRole(ctx, keyVaultID, cryptoOfficerRoleDefinitionID, testUserPrincipalID)
+	if err != nil {
+		return fmt.Errorf("failed to assign Crypto Officer role to test user: %w", err)
+	}
+	err = env.RBACManager.EnsureRole(ctx, keyVaultID, keyVaultAdminRoleDefinitionID, testUserPrincipalID)
+	if err != nil {
+		return fmt.Errorf("failed to assign Administrator role to test user: %w", err)
+	}
+
+	// RBAC propagation is handled by retry logic in the clients (403 errors are retried)
+
+	return nil
+}
+
+// createKeyVaultKey creates a key in the Key Vault
+func (env *Environment) createKeyVaultKey(ctx context.Context, keyVaultName, keyName string, cred *azidentity.DefaultAzureCredential) (*azkeys.KeyBundle, error) {
+	// Add retry options for Key Vault operations that may encounter RBAC propagation delays
+	// RBAC assignments can take time to propagate, resulting in 403 Forbidden errors
+	// With 15 retries at 5 second intervals = 75 seconds total retry time
+	keyClientOptions := &azkeys.ClientOptions{
+		ClientOptions: policy.ClientOptions{
+			Retry: policy.RetryOptions{
+				MaxRetries: 15,
+				RetryDelay: time.Second * 5,
+				StatusCodes: []int{
+					http.StatusForbidden, // RBAC assignments haven't propagated to Key Vault yet
+				},
+			},
+		},
+	}
+
+	keyClient, err := azkeys.NewClient(fmt.Sprintf("https://%s.vault.azure.net/", keyVaultName), cred, keyClientOptions)
```

Comment:
IIRC the KV should have this endpoint on the GET response - you could pass the whole KV object in and might be able to just read it off there?